File: Tutorial4-2Subcomponents/src/main/java/com/example/tutorial4_2subcomponents/di/ActivityModule.java
package com.example.tutorial4_2subcomponents.di;

import com.example.tutorial4_2subcomponents.MainActivity;
import com.example.tutorial4_2subcomponents.ToastMaker;

import dagger.Module;
import dagger.Provides;

@Module
public class ActivityModule {

    private final MainActivity mainActivity;

    public ActivityModule(MainActivity mainActivity) {
        this.mainActivity = mainActivity;
    }

    @ActivityScope
    @Provides
    ToastMaker provideToastMaker() {
        return  new ToastMaker(mainActivity);
    };
}

File: Tutorial4-2Subcomponents/src/main/java/com/example/tutorial4_2subcomponents/di/ActivityModule.kt
package com.example.tutorial4_2subcomponents.di

import com.example.tutorial4_2subcomponents.MainActivity
import com.example.tutorial4_2subcomponents.model.ToastMaker
import dagger.Module
import dagger.Provides

/**
 * This is the module of sub-component [ActivitySubComponent]
 */
@Module
class ActivityModule(private val mainActivity: MainActivity) {

    @ActivityScope
    @Provides
    fun provideToastMaker(): ToastMaker {
        return ToastMaker(mainActivity)
    }

}